src/order/complete_lattice.lean

```diff
@@ -382,113 +390,201 @@ end
+  (λ (h : Sup s = ⊤) b hb, by rwa [←h, lt_Sup_iff] at hb)
+  (λ h, top_unique $ le_of_not_gt $ λ h',
     let ⟨a, ha, h⟩ := h _ h' in
     lt_irrefl a $ lt_of_le_of_lt (le_Sup ha) h)
 
 lemma Inf_eq_bot : Inf s = ⊥ ↔ (∀b>⊥, ∃a∈s, a < b) :=
 @Sup_eq_top (order_dual α) _ _
 
-lemma lt_supr_iff {f : ι → α} : a < supr f ↔ (∃i, a < f i) :=
-lt_Sup_iff.trans exists_range_iff
-
-lemma infi_lt_iff {f : ι → α} : infi f < a ↔ (∃i, f i < a) :=
-Inf_lt_iff.trans exists_range_iff
+lemma lt_supr_iff {f : ι → α} : a < supr f ↔ ∃ i, a < f i := lt_Sup_iff.trans exists_range_iff
+lemma infi_lt_iff {f : ι → α} : infi f < a ↔ ∃ i, f i < a := Inf_lt_iff.trans exists_range_iff
 
 end complete_linear_order
 
 /-
 ### supr & infi
 -/
 
+section has_Sup
+variables [has_Sup α] {f g : ι → α}
+
+lemma Sup_range : Sup (range f) = supr f := rfl
+lemma Sup_eq_supr' (s : set α) : Sup s = ⨆ a : s, a := by rw [supr, subtype.range_coe]
+
+lemma supr_congr (h : ∀ i, f i = g i) : (⨆ i, f i) = ⨆ i, g i := congr_arg _ $ funext h
+
+lemma function.surjective.supr_comp {f : ι → ι'} (hf : surjective f) (g : ι' → α) :
+  (⨆ x, g (f x)) = ⨆ y, g y :=
+by simp only [supr, hf.range_comp]
+
+protected lemma function.surjective.supr_congr {g : ι' → α} (h : ι → ι') (h1 : surjective h)
+  (h2 : ∀ x, g (h x) = f x) :
+  (⨆ x, f x) = ⨆ y, g y :=
+by { convert h1.supr_comp g, exact (funext h2).symm }
+
+@[congr] lemma supr_congr_Prop {p q : Prop} {f₁ : p → α} {f₂ : q → α} (pq : p ↔ q)
+  (f : ∀ x, f₁ (pq.mpr x) = f₂ x) : supr f₁ = supr f₂ :=
+begin
+  have := propext pq, subst this,
+  congr' with x,
+  apply f
+end
+
+lemma supr_range' (g : β → α) (f : ι → β) : (⨆ b : range f, g b) = ⨆ i, g (f i) :=
+by rw [supr, supr, ← image_eq_range, ← range_comp]
+
+lemma Sup_image' {s : set β} {f : β → α} : Sup (f '' s) = ⨆ a : s, f a :=
+by rw [supr, image_eq_range]
+
+end has_Sup
+
+section has_Inf
+variables [has_Inf α] {f g : ι → α}
+
+lemma Inf_range : Inf (range f) = infi f := rfl
+lemma Inf_eq_infi' (s : set α) : Inf s = ⨅ a : s, a := @Sup_eq_supr' (order_dual α) _ _
+
+lemma infi_congr (h : ∀ i, f i = g i) : (⨅ i, f i) = ⨅ i, g i := congr_arg _ $ funext h
+
+lemma function.surjective.infi_comp {f : ι → ι'} (hf : surjective f) (g : ι' → α) :
+  (⨅ x, g (f x)) = ⨅ y, g y :=
+@function.surjective.supr_comp (order_dual α) _ _  _ f hf g
+
+lemma function.surjective.infi_congr {g : ι' → α} (h : ι → ι') (h1 : surjective h)
+  (h2 : ∀ x, g (h x) = f x) : (⨅ x, f x) = ⨅ y, g y :=
+@function.surjective.supr_congr (order_dual α) _ _ _ _ _ h h1 h2
+
+@[congr]lemma infi_congr_Prop {p q : Prop} {f₁ : p → α} {f₂ : q → α}
+  (pq : p ↔ q) (f : ∀ x, f₁ (pq.mpr x) = f₂ x) : infi f₁ = infi f₂ :=
+@supr_congr_Prop (order_dual α) _ p q f₁ f₂ pq f
+
+lemma infi_range' (g : β → α) (f : ι → β) : (⨅ b : range f, g b) = ⨅ i, g (f i) :=
+@supr_range' (order_dual α) _ _ _ _ _
+
+lemma Inf_image' {s : set β} {f : β → α} : Inf (f '' s) = (⨅ a : s, f a) :=
+@Sup_image' (order_dual α) _ _ _ _
+
+end has_Inf
```

Comment:
Where did these move from? Would you be able to do this move in a spin-off PR?

Reply:
They moved from lower down the file. I fear that splitting this PR will make it rot horribly.

Reply:
Can you make sure to summarize this move in the PR description? I don't see a mention of it.